screen-orientation/onchange-event.html

```diff
@@ -1,80 +1,37 @@
 <!DOCTYPE html>
-<html>
-<body>
 <script src="/resources/testharness.js"></script>
 <script src="/resources/testharnessreport.js"></script>
 <script>
-
-var changeTest = async_test("Test that orientationchange event is fired when the orientation changes.");
-var noChangeTest = async_test("Test that orientationchange event is not fired when the orientation does not change.");
-
-var orientations = [
-  'portrait-primary',
-  'portrait-secondary',
-  'landscape-primary',
-  'landscape-secondary'
-];
-
-var currentIndex = orientations.indexOf(window.screen.orientation.type);
-// Count the number of calls received from the EventHandler set on screen.orientation.onchange.
-var orientationChangeEventHandlerCalls = 0;
-// Count the number of calls received from the listener set with screen.orientation.addEventListene().
-var orientationChangeEventListenerCalls = 0;
-
-var orientationChangeContinuation = null;
-
-function getNextIndex() {
-  return (currentIndex + 1) % orientations.length;
-}
-
-window.screen.orientation.onchange = function() {
-  orientationChangeEventHandlerCalls++;
-  if (orientationChangeEventContinuation) {
-    setTimeout(orientationChangeEventContinuation);
-    orientationChangeEventContinuation = null;
-  }
-};
-
-window.screen.orientation.addEventListener('change', function() {
-  orientationChangeEventListenerCalls++;
-});
-
-function runNoChangeTest() {
-  screen.orientation.lock(orientations[currentIndex]).then(function() {}, function() {});
-
-  noChangeTest.step(function() {
-    assert_equals(screen.orientation.type, orientations[currentIndex]);
-    assert_equals(orientationChangeEventHandlerCalls, orientations.length);
-    assert_equals(orientationChangeEventListenerCalls, orientations.length);
-  });
-
-  noChangeTest.done();
-}
-
-var i = 0;
-function runChangeTest() {
-  screen.orientation.lock(orientations[getNextIndex()]).then(function() {}, function() {});
-  currentIndex = getNextIndex();
-
-  orientationChangeEventContinuation = function() {
-    changeTest.step(function() {
-      assert_equals(screen.orientation.type, orientations[currentIndex]);
-      assert_equals(orientationChangeEventHandlerCalls, i + 1);
-      assert_equals(orientationChangeEventListenerCalls, i + 1);
-    });
-
-    ++i;
-    if (i == orientations.length) {
-      changeTest.done();
-      runNoChangeTest();
-    } else {
-      runChangeTest();
+promise_test(t => {
+  const type = screen.orientation.type;
+  // wrap onchange listener in new promise to catch unreached assertion
+  return new Promise(async (resolve, reject) => {
+    screen.orientation.onchange = () => { 
+      reject(new Error("change event should not be fired"));
     }
-  };
-}
-
-runChangeTest();
+    await screen.orientation.lock(type);
+    assert_equals(screen.orientation.type, type);
+    setTimeout(resolve, 1000);
```

Comment:
Without a timeout, the Promise will resolve immediately once the above assertion is completed, even though the `onchange` is fired.
 I didn't find a good way more than using a `setTimeout` :(

Reply:
but that's ok, I think: 

1. if `await screen.orientation.lock(type);` invokes the event handler, then we know the test has failed. 
1. resolve();

Reply:
(i.e., we can safely drop setTimeout())

Reply:
I've tested, if drop `setTimeout()`, the test will still pass even the event is fired. That means the `resolve();` will be invoked before event handler.

Reply:
Interesting! Which browser is that in?

Reply:
Chrome 68 on Android

Reply:
@mounirlamouri is that a bug in Chrome? ☝️☝️☝️

Reply:
We queue the event when we dispatch it so I guess it ends up being asynchronous while the promise may not be. It might be what the spec used to say. I can't remember :)

If this is passing in another browser, I would suggest to drop the `setTimeout()` and file a bug against Chrome. If it is not, I would suggest to change `setTimeout(resolve, 100);` to `setTimeout(resolve);` and add a test specifically for the race between event and promise and have Chrome fail it.

Reply:
Here's an another example to use timeout to detect a misfire event https://github.com/w3c/web-platform-tests/blob/master/mediacapture-streams/MediaStream-idl.https.html#L52.

It's the same result in Fifefox 60 Android, will use `setTimeout(resolve)` in new commit.

>  add a test specifically for the race between event and promise and have Chrome fail it.

I'm not quite understand, is this new test specific for testing race between event and promise, and irrelevant to screen orientation?

Reply:
I meant to test the race between both in the context of the orientation change event.

Reply:
I think we should just follow the spec. Just test that the event doesn't immediately fire. Events randomly firing at strange times is not something we can reliably test... consider, we set timeout for 1000 and the event fires at 1001.